### PR TITLE
fix(multipart upload): lock down multipart upload

### PR DESCRIPTION
**Pull request by ctrlc03**
Added checks to prevent overwriting of files with multi part upload and that the correct zKey index is uploaded

### CODE REVIEW

File: packages/actions/test/unit/security.test.ts [hunk 1]
```diff
@@ -188,6 +260,115 @@ describe("Security", () => {
         })
     })
 
+    if (envType === TestingEnvironment.PRODUCTION) {
+        // Tests related to multi part upload security
+        // @note We want to make sure that a contributor can
+        // 1. only upload when in contributing status (current contributor)
+        // 2. only upload a file with the correct name
+        // 3. only upload a valid zkey file and previous zkey
```
Comment (0xjei) [inline packages/actions/test/unit/security.test.ts:268]: Why `...and previous zKey`?

File: packages/backend/src/functions/storage.ts [hunk 2]
```diff
@@ -53,6 +53,71 @@ const checkPreConditionForCurrentContributorToInteractWithMultiPartUpload = asyn
         logAndThrowError(SPECIFIC_ERRORS.SE_STORAGE_CANNOT_INTERACT_WITH_MULTI_PART_UPLOAD)
 }
 
+/**
+ * Helper function to check whether a contributor is uploading a file related to its contribution.
+ * @param contributorId <string> - the unique identifier of the contributor.
+ * @param ceremonyId <string> - the unique identifier of the ceremony.
+ * @param objectKey <string> - the object key of the file being uploaded.
+ */
+const checkUploadingFileValidity = async (contributorId: string, ceremonyId: string, objectKey: string) => {
```
Comment (0xjei) [inline packages/backend/src/functions/storage.ts:62]: Wow, looks really good!

File: packages/backend/src/functions/storage.ts [hunk 3]
```diff
@@ -53,6 +53,71 @@ const checkPreConditionForCurrentContributorToInteractWithMultiPartUpload = asyn
...
+        }
+    } else logAndThrowError(SPECIFIC_ERRORS.SE_STORAGE_CANNOT_INTERACT_WITH_MULTI_PART_UPLOAD)
+}
+
+/**
+ * Helper function that confirms whether a bucket is used for a ceremony.
+ * @dev this helps to prevent unauthorized access to coordinator's buckets.
+ * @param bucketName
+ */
+const checkIfBucketIsDedicatedToCeremony = async (bucketName: string) => {
```
Comment (0xjei) [inline packages/backend/src/functions/storage.ts:105]: cute!

**Review by 0xjei — APPROVED** [review]
LGTM :fire: Great job!

File: packages/actions/test/unit/security.test.ts (hunk 1 above)
Comment (ctrlc03) [reply packages/actions/test/unit/security.test.ts:268]: had the wrong assumption when started designing this and never removed it from the comments🙃

**Pull request merged**